### PR TITLE
(MODULES-11255) Add basic tasks to manage packages

### PR DESCRIPTION
**Pull request by smortex**
Add a task to list current packages:

```
C:\> bolt task run chocolatey::status -t localhost
Started on localhost...
Finished on localhost:
{"status":[{"package":"chocolatey","version":"0.11.3"},{"package":"puppet-bolt","version":"3.20.0"}]}
Successful on 1 target: localhost
Ran on 1 target in 2.6 sec
C:\>
```

Add tasks to install, upgrade and uninstall packages:

```
C:\> bolt task run chocolatey action=install package=pdk version=2.0.0.0 -t localhost
C:\> bolt task run chocolatey action=upgrade package=pdk -t localhost
C:\> bolt task run chocolatey action=uninstall package=pdk -t localhost
```

Add a task to pin/unpin packages:

```
C:\> bolt task run chocolatey::pin action=add package=puppet-bolt -t localhost
C:\> bolt task run chocolatey::pin action=list -t localhost
Started on localhost...
Finished on localhost:
{"status":[{"package":"puppet-bolt","version":"3.20.0"}]}
Successful on 1 target: localhost
Ran on 1 target in 2.6 sec
C:\> bolt task run chocolatey::pin action=remove package=pdk -t localhost
```

Add a task to list outdated packages:

```
C:\> bolt task run chocolatey::outdated -t localhost
Started on localhost...
Finished on localhost:
{"status":[{"package":"puppet-bolt","version":"3.20.0","available_version":"3.21.0","pinned":false}]}
Successful on 1 target: localhost
Ran on 1 target in 10.14 sec
C:\>
```

### CODE REVIEW

**Comment by smortex** [comment]
This is currently a draft to allow the community to provide feedback on the interface of the tasks.  Cc @neomilium, @bastelfreak, @mtancoigne :grin:

**Review by bastelfreak — DISMISSED** [review]

File: tasks/outdated.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  output, status = Open3.capture2('choco', 'outdated', '--no-color')
+  raise Puppet::Error, output unless [0, 2].include?(status.exitstatus)
+
+  result = []
+
+  output.split("\n").each do |line|
+    parts = line.split('|')
+
+    next if parts.count != 4
+    next if parts[0].include?(' ')
```
Comment (smortex) [inline tasks/outdated.rb:17]: @neomilium told me about `--limit-output` which avoid the need for this.

File: tasks/status.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  output, status = Open3.capture2('choco', 'list', '--local-only', '--no-color')
+  raise Puppet::Error, output if status != 0
+
+  result = []
+
+  output.split("\n").each do |line|
+    parts = line.split(' ')
+
+    next if parts.count != 2
+    next if parts[0] == 'Chocolatey' # package name is "chocolatey", "Chocolatey" is the header line
```
Comment (smortex) [inline tasks/status.rb:17]: Another location `--limit-output` should help.

File: tasks/status.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      version: parts[1],
+    }
+  end
+
+  puts result.to_json
+  exit 0
+rescue Puppet::Error => e
+  puts({ status: 'failure', error: e.message }.to_json)
+  exit 1
+end
```
Comment (smortex) [inline tasks/status.rb:30]: @neomilium suggests looking into the puppetlabs/ruby_task_helper module for convenient helpers.

**Review by bastelfreak — DISMISSED** [review]

**Comment by michaeltlombardi** [comment]
Given that this is a Windows-only module and task, is there a reason to write it to require the Puppet Agent/Ruby instead of using PowerShell?

**Comment by smortex** [comment]
@michaeltlombardi as a POSIX system user who does a lot of Ruby and has to manage a few windows nodes, I have not even *though* about using powershell :rofl: I jumped directly on the tools I am used to and master.

**Comment by michaeltlombardi** [comment]
I only ask because the primary drawback is that using ruby means these tasks can _only_ be executed against nodes with the agent installed. Not a _huge_ deal just wanted to check!

**Review by bastelfreak — DISMISSED** [review]

File: .github/workflows/spec.yml [hunk 4]
```diff
@@ -120,6 +120,10 @@ jobs:
           buildevents cmd $TRACE_ID $STEP_ID 'bundle env' -- bundle env
           echo ::endgroup::
 
+      - name: Create task helper symlink
+        run: |
+          ln -s "${PWD}/spec/fixtures/modules/ruby_task_helper" ..
+
```
Comment (smortex) [inline .github/workflows/spec.yml:126]: FYI, I created https://github.com/puppetlabs/pdk-templates/pull/461 to handle this in the future without regression.

**Review by bastelfreak — DISMISSED** [review]

File: .github/workflows/spec.yml (hunk 4 above)
Comment (smortex) [reply .github/workflows/spec.yml:126]: Merged. I updated `.sync.yml` to enable the inclusion of this code.

File: tasks/status.rb (hunk 3 above)
Comment (smortex) [reply tasks/status.rb]: Done

**Comment by smortex** [comment]
I think this has settled so I squashed the commits and updated the PR description accordingly.  I have been using these tasks for some time now and consider this ready for merging :tada:

**Review by bastelfreak — APPROVED** [review]

**Pull request merged**